onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,199 @@
+```
+
+When the inputs of Gradient are the tensors named in "xs", the computation
+can be optimized. More specifically, intermediate variables in forward pass can
+be reused if the gradient is computed via reverse-mode auto-differentiation.
+
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Gradient,
+    12,
+    OpSchema()
+        .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+        .SetDomain("ai.onnx.training")
+        .SetDoc(Gradient_ver12_doc)
+        .Input(
+            0,
+            "Inputs",
+            "The values fed into graph identified by the attributes. "
+            "The i-th input is the value of the i-th tensor specified in the "
+            "concatenated list of attribute \"xs\" and attribute \"zs\". For "
+            "example, if xs=[\"A\", \"B\"] and zs=[\"C\"], the first input is "
+            "used as the value of symbol \"A\" and the 3rd input is "
+            "substituted for all the occurrences of \"C\".",
+            "T1",
+            OpSchema::Variadic,
+            false)
+        .Output(
+            0,
+            "Outputs",
+            "The gradient of the tensors specified in attribute \"y\" "
+            "with respect to \"xs\". The i-th output is the gradient "
+            "of \"y\" with respect to the i-th tensor specified in the "
+            "attribute \"xs\".",
+            "T2",
+            OpSchema::Variadic,
+            false)
+        .Attr(
+            "xs",
+            "Input tensor names of the differentiated sub-graph. It "
+            "contains and only contains the necessary differentiated "
+            "inputs of a (sub-)graph. Variables (usually called "
+            "intermediate variables) which can be generated by inputs "
+            "cannot be included in this attribute.",
+            AttributeProto::STRINGS)
+        .Attr(
+            "zs",
+            "Input tensor names of the differentiated sub-graph. It "
+            "contains and only contains the necessary non-differentiated "
+            " inputs of a (sub-)graph. Variables (usually called "
+            "intermediate variables) which can be generated by inputs "
+            "cannot be included in this attribute.",
+            AttributeProto::STRINGS,
+            OPTIONAL)
+        .Attr(
+            "y",
+            "The targeted tensor. It can be viewed as the output of the "
+            "differentiated function. The attribute \"xs\" and attribute "
+            "\"zs\" are the minimal independent variable set that determines "
+            "the value of \"y\".",
+            AttributeProto::STRING)
+        .TypeConstraint(
+            "T1",
+            OpSchema::all_tensor_types(),
+            "Allow outputs to be any kinds of tensors.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(float16)",
+             "tensor(float)",
+             "tensor(double)"},
+            "Allow inputs to be any kind of tensor."));
```

Comment:
@wschin Perhaps "any kind of differentiable tensor type"?

Reply:
I will change it to `Allow inputs to be any kind of floating-point tensor`. Maybe with some advanced math techniques, other types are differentiable too.